base_layer/wallet/src/output_manager_service/service.rs

```diff
@@ -450,46 +450,120 @@ where
+                // max_utxo_query_size
+                let rounds =
+                    ((unspent_outputs.len() as f32) / (self.config.max_utxo_query_size as f32 + 0.1)) as usize + 1;
 
-                let service_request = BaseNodeProto::BaseNodeServiceRequest {
-                    request_key,
-                    request: Some(request),
-                };
-                // TODO Remove this once this bug is fixed
-                trace!(target: LOG_TARGET, "About to attempt to send query to base node");
-                self.outbound_message_service
-                    .send_direct(
-                        pk.clone(),
-                        OutboundEncryption::None,
-                        OutboundDomainMessage::new(TariMessageType::BaseNodeRequest, service_request),
-                    )
-                    .await?;
-                // TODO Remove this once this bug is fixed
-                trace!(target: LOG_TARGET, "Query sent to Base Node");
-                self.pending_utxo_query_keys.insert(request_key, output_hashes);
-                let state_timeout = StateDelay::new(self.config.base_node_query_timeout, request_key);
-                utxo_query_timeout_futures.push(state_timeout.delay().boxed());
-                debug!(
-                    target: LOG_TARGET,
-                    "Output Manager Sync query ({}) sent to Base Node", request_key
-                );
-                Ok(request_key)
+                for r in 0..rounds {
+                    let mut output_hashes = Vec::new();
+                    for uo_hash in
+                        unspent_outputs.drain(..cmp::min(self.config.max_utxo_query_size, unspent_outputs.len()))
+                    {
+                        output_hashes.push(uo_hash);
+                    }
+                    let request_key = OsRng.next_u64();
+                    if first_request_key == 0 {
+                        first_request_key = request_key;
+                    }
+
+                    let request = BaseNodeRequestProto::FetchUtxos(BaseNodeProto::HashOutputs {
+                        outputs: output_hashes.clone(),
+                    });
+
+                    let service_request = BaseNodeProto::BaseNodeServiceRequest {
+                        request_key,
+                        request: Some(request),
+                    };
+
+                    let send_message_response = self
+                        .outbound_message_service
+                        .send_direct(
+                            pk.clone(),
+                            OutboundEncryption::None,
+                            OutboundDomainMessage::new(TariMessageType::BaseNodeRequest, service_request),
+                        )
+                        .await?;
+
+                    // Here we are going to spawn a non-blocking task that will monitor and log the progress of the
+                    // send process.
+                    let owned_request_key = request_key;
```

Comment:
`u64` is a `Copy` type so you can just use it in the async block

Reply:
You can't, it complained that it was dropped too early when I used it below.

Reply:
Huh, you are right. I must of left out the `move` when I first wrote this.